contracts/examples/erc1155-marketplace/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+			"auction has not ended yet!"
+		);
+
+		self.clear_auction_for_token(&type_id, &token_id);
+
+		if auction.current_winner != Address::zero() {
+			let percentage_cut = self.get_percentage_cut();
+			let cut_amount = self.calculate_cut_amount(&auction.current_bid, percentage_cut);
+			let amount_to_send = &auction.current_bid - &cut_amount;
+
+			let mut claimable_funds = self.get_claimable_funds();
+			claimable_funds += &cut_amount;
+			self.set_claimable_funds(&claimable_funds);
+
+			// send part of the bid to the original owner
+			self.send().direct_egld(&auction.original_owner, &amount_to_send, b"sold token");
+
+			// send token to winner
+			self.asnyc_transfer_token(&type_id, &token_id, &auction.current_winner);
+		} else {
+			// return to original owner
+			self.asnyc_transfer_token(&type_id, &token_id, &auction.original_owner);
+		}
+
+		Ok(())
+	}
+
+	// views
+
+	#[view(isUpForAuction)]
+	fn is_up_for_auction(&self, type_id: &BigUint, token_id: &BigUint) -> bool {
+		!self.is_empty_auction_for_token(type_id, token_id)
+	}
+
+	#[view(getAuctionStatus)]
+	fn get_auction_status(
+		&self,
+		type_id: BigUint,
+		token_id: BigUint,
+	) -> SCResult<Auction<BigUint>> {
+		require!(
+			self.is_up_for_auction(&type_id, &token_id),
+			"Token is not up for auction"
+		);
+
+		Ok(self.get_auction_for_token(&type_id, &token_id))
+	}
+
+	#[view(getCurrentWinningBid)]
+	fn get_current_winning_bid(&self, type_id: BigUint, token_id: BigUint) -> SCResult<BigUint> {
+		require!(
+			self.is_up_for_auction(&type_id, &token_id),
+			"Token is not up for auction"
+		);
+
+		Ok(self.get_auction_for_token(&type_id, &token_id).current_bid)
+	}
+
+	#[view(getCurrentWinner)]
+	fn get_current_winner(&self, type_id: BigUint, token_id: BigUint) -> SCResult<Address> {
+		require!(
+			self.is_up_for_auction(&type_id, &token_id),
+			"Token is not up for auction"
+		);
+
+		Ok(self
+			.get_auction_for_token(&type_id, &token_id)
+			.current_winner)
+	}
+
+	// private
+
+	fn try_create_auction(
+		&self,
+		type_id: &BigUint,
+		token_id: &BigUint,
+		original_owner: &Address,
+		min_bid: &BigUint,
+		max_bid: &BigUint,
+		deadline: u64,
+	) -> SCResult<()> {
+		require!(
+			!self.is_up_for_auction(&type_id, &token_id),
+			"There is already an auction for that token"
+		);
+		require!(
+			min_bid > &0 && min_bid <= max_bid,
+			"Min bid can't be 0 or higher than max bid"
+		);
+		require!(
+			deadline > self.get_block_timestamp(),
+			"Deadline can't be in the past"
+		);
+
+		self.set_auction_for_token(
+			&type_id,
+			&token_id,
+			&Auction {
+				min_bid: min_bid.clone(),
+				max_bid: max_bid.clone(),
+				deadline,
+				original_owner: original_owner.clone(),
+				current_bid: BigUint::zero(),
+				current_winner: Address::zero(),
+			},
+		);
+
+		Ok(())
+	}
+
+	// TODO: Replace with Proxy in the next framework version
+	fn asnyc_transfer_token(&self, type_id: &BigUint, token_id: &BigUint, to: &Address) {
```

Comment:
is this ESDT transfer ?

Reply:
or this is the actual call for the ERC-1155 contract to transfer the ownership ?

Reply:
Yep, it's a call to transfer ownership of the ERC-1155 token. The marketplace contract will own the token until it's sold. I initially wanted to make use of the approval system for this, but I thought it would be pretty unsafe. Couldn't really find a solution to the case where the owner of the token revokes the approval and the marketplace SC loses its right to transfer the token.

Reply:
sounds fair.